### PR TITLE
[Attention] Add UltraQuant 4-bit KV cache backend (FlyDSL D=256)

### PR DESCRIPTION
**Pull request by aditi-amd**
## Purpose
Adds an `ultraquant_4bit` KV-cache dtype for long-context / agentic serving, where the KV cache is the limiter. UltraQuant stores the cache at 4 bits, FP4 E2M1 codes with UE8M0 (power-of-two) per-group-32 scales, and Hadamard-rotated keys, giving **~2× KV density over fp8**, so more concurrency fits before eviction and decode stays memory-efficient under load.
Decode runs on a FlyDSL `D=256` kernel for GQA group sizes **{6, 8, 16}**, with a Triton unified-attention fallback and numerical reference paths.

## Design
- New `UltraQuantAttentionImpl`, a standalone subclass of `TurboQuantAttentionImpl`. 
- `TurboQuantAttentionBackend.get_impl_cls()` dispatches to the UQ impl on the `ultraquant_4bit` cache dtype.
- New ops under `vllm/v1/attention/ops/ultraquant/` (FP4/UE8M0 format, Hadamard store, dequant, Triton unified attention, reference) and the FlyDSL D=256 decode kernel/launcher.

## Test Plan
- **Unit tests** (`tests/kernels/ultraquant/`): store, dequant, decode, and FlyDSL-vs-reference numerical checks, plus config.
- **Serve smokes**: Qwen3.8 (GQA-16, TP=8) and Qwen3.6-27B (GQA-6, TP=2), confirming FlyDSL dispatch (no Triton fallback) and coherent output.
- **Accuracy** (KV format is the only variable; sampling fixed): GPQA-Diamond.
- **Performance**: SemiAnalysis-style 262K-token agentic replay, throughput/latency and peak KV utilization across concurrency.

## Usage
Enable the 4-bit KV cache with `--kv-cache-dtype ultraquant_4bit` (D=256, GQA {6,8,16}):

vllm serve /path/to/Qwen3.8-2.4T-A95B-MXFP4 \
  --tensor-parallel-size 8 \
  --attention-backend ROCM_AITER_UNIFIED_ATTN \
  --kv-cache-dtype ultraquant_4bit \
  --max-model-len 163840 \
  --reasoning-parser qwen3 \
  --trust-remote-code

The 4-bit KV cache composes with MXFP4 weights. Decode auto-dispatches to the FlyDSL D=256 kernel (gfx950), otherwise it falls back to Triton unified attention. All accuracy/perf runs use the sampling settings recommended in the model's Hugging Face card (Qwen3.8: `temperature=1.0, top_p=0.95, top_k=20`).

## Test Results
**Accuracy — GPQA-Diamond, mean over multiple seeds.**
UltraQuant is at parity with the fp8 (KV8) baseline — in fact marginally ahead on the mean. The decode kernel is numerically exact, so per-seed differences are sampling variance (the arms trade places seed to seed).
| Model | UltraQuant 4-bit (mean) | KV8 fp8 (mean) |
|---|---|---|
| Qwen3.8 | **93.27%** | 92.42% |
| Qwen3.6-27B | **86.87%** | 85.61% |

**Performance: throughput–latency Pareto (262K-token requests, MI355X):**
Qwen3.8 (GQA-16, TP=8). UQ and KV8 track together at low concurrency; UQ pulls ahead from **C=32** onward. KV8 saturates its cache at **C=38 (100%, evicting)** and throughput collapses, while UQ is still at ~44% cache and keeps scaling.

sa_pareto_qwen38_c8_42_x160

| Concurrency | UQ tok/s | KV8 tok/s | Δ | UQ peak KV | KV8 peak KV |
|---:|---:|---:|---:|---:|---:|
| 8  | 193.9 | 199.4 | −2.7%  | 8.2%  | 14.1% |
| 16 | 280.2 | 288.0 | −2.7%  | 17.4% | 32.0% |
| 32 | 374.0 | 363.1 | +3.0%  | 35.1% | 66.3% |
| 36 | 398.9 | 299.6 | +33.1% | 39.7% | 82.4% |
| 38 | 414.6 | 225.4 | +84.0% | 44.2% | **100.0%** |
| 42 | 430.7 | 114.7 | +275%  | 45.4% | 99.9% |
---

UQ first wins at **C=32 before any eviction** (KV8 only 66% cache), and the gap widens sharply once KV8 hits the eviction wall at C=38.
Qwen3.6-27B (GQA-6, TP=2). Same shape; UQ wins even before KV8 evicts at C=42 UQ delivers +9% throughput at lower ITL while KV8 is only ~39% cache. UQ peak KV is ~half of KV8 at every point.

sa_pareto_qwen36_tp2_x70

| Concurrency | UQ tok/s | KV8 tok/s | Δ | UQ peak KV | KV8 peak KV |
|---:|---:|---:|---:|---:|---:|
| 8  | 268.3 | 271.8 | −1.3% | 4.8%  | 8.4%  |
| 16 | 383.3 | 391.4 | −2.1% | 9.7%  | 15.2% |
| 32 | 519.9 | 529.5 | −1.8% | 17.6% | 32.1% |
| 34 | 545.1 | 537.6 | +1.4% | 19.0% | 31.8% |
| 42 | 592.0 | 543.1 | +9.0% | 22.6% | 38.7% |

## Notes
- Requires a FlyDSL-capable ROCm build (gfx950) for the D=256 decode kernel; falls back to Triton unified attention otherwise.
- Benchmark figures were measured across differing vLLM / AITER / FlyDSL bases, so absolute throughput/latency (and the 8-bit baseline) are indicative rather than bit-identical across environments; the accuracy and relative UQ-vs-KV8 trends are the takeaway.

## References
[UltraQuant 4bit KV](https://arxiv.org/pdf/2606.20474)

### CODE REVIEW

File: vllm/v1/attention/backends/turboquant_attn.py
Comment (BowenBao) [inline vllm/v1/attention/backends/turboquant_attn.py:1]: Seems like ultraquant depends on turboquant, is that a hard requirement?

File: vllm/v1/attention/ops/flydsl_kernels/ultraquant_decode_hd256.py [hunk 1]
```diff
@@ -0,0 +1,983 @@
...
+                    # ---- load this lane's 32 raw-Q head-dims (4 x dwordx4) ----
+                    _v = []
+                    for _j in range_constexpr(4):
+                        _raw = buffer_ops.buffer_load(
+                            q_rsrc,
+                            (_qb + fx.Int32(_j * 8)) // fx.Int32(2),
+                            vec_width=4,
+                            dtype=T.i32,
+                        )
+                        _rb = vector.bitcast(T.vec(8, T.bf16), _raw)
```
Comment (BowenBao) [inline vllm/v1/attention/ops/flydsl_kernels/ultraquant_decode_hd256.py:400]: this assumes and hardcodes query to bf16. however in https://github.com/aditi-amd/vllm/blob/229d47dbf9420118b75ce96eafa4a8372d0b1954/vllm/v1/attention/backends/turboquant_attn.py#L134 float16 is marked as supported.

File: vllm/v1/attention/ops/ultraquant/triton_store.py [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+    )
+
+
+_SUPPORTED_DTYPES = {torch.float16, torch.bfloat16}
+
+# Triton only lets a @jit body reach a module global if it is a `constexpr`
+# instance, so the midpoint table is wrapped here and indexed via `.value`
+# inside the unrolled `static_range` (the index is a Python int at trace time,
+# so each comparison lowers to an immediate).
+_MIDPOINTS_C = tl.constexpr(tuple(float(m) for m in MIDPOINTS_SORTED))
```
Comment (BowenBao) [inline vllm/v1/attention/ops/ultraquant/triton_store.py]: not sure but this might be a problem for envs without triton, calling `tl.constexpr(...)` at import time

File: vllm/v1/attention/backends/turboquant_attn.py [hunk 3]
```diff
@@ -162,18 +173,54 @@ def supported_kv_cache_layouts(cls) -> tuple[KVCacheLayout, ...]:
...
+            return 64
+        return super().get_preferred_block_size(default_block_size)
 
     @classmethod
     def supports_attn_type(cls, attn_type: str) -> bool:
         return attn_type == AttentionType.DECODER
 
+    @classmethod
+    def supports_sink(cls) -> bool:
+        return True
```
Comment (BowenBao) [inline vllm/v1/attention/backends/turboquant_attn.py:201]: is this intended?

File: vllm/engine/arg_utils.py [hunk 4]
```diff
@@ -2122,16 +2122,23 @@ def create_engine_config(
...
             boundary = TurboQuantConfig.get_boundary_skip_layers(model_config)
             existing = set(cache_config.kv_cache_dtype_skip_layers)
+            combined = existing | set(boundary)
+            numeric_values = {value for value in combined if value.isdigit()}
+            special_values = combined - numeric_values
             cache_config.kv_cache_dtype_skip_layers = sorted(
-                existing | set(boundary), key=int
-            )
+                numeric_values, key=int
+            ) + sorted(special_values)
```
Comment (BowenBao) [inline vllm/engine/arg_utils.py]: is this for a particular bugfix?

**Review by BowenBao — COMMENTED** [review]
Thanks, a few comments as going through the PR

File: vllm/v1/attention/backends/turboquant_attn.py
Comment (aditi-amd) [reply vllm/v1/attention/backends/turboquant_attn.py:1]: UltraQuant doesn't functionally depend on TurboQuant, however to they do share some common code structures, shared prefill, cache-layout machinery both bring 4-bit.

File: vllm/v1/attention/backends/turboquant_attn.py (hunk 3 above)
Comment (aditi-amd) [reply vllm/v1/attention/backends/turboquant_attn.py:201]: The flydsl UQ path doesn't support sinks but the fallback triton does have the sink support and that is why it was added.

File: vllm/v1/attention/ops/flydsl_kernels/ultraquant_decode_hd256.py (hunk 1 above)
Comment (aditi-amd) [reply vllm/v1/attention/ops/flydsl_kernels/ultraquant_decode_hd256.py:400]: Good catch! UQ is only doing bf16 Q and the models we tested were also all bf16. Delivered the fix to remove fp16 as supported.

File: vllm/v1/attention/ops/ultraquant/triton_store.py (hunk 2 above)
Comment (aditi-amd) [reply vllm/v1/attention/ops/ultraquant/triton_store.py]: Good suggestion, i have fixed it.

File: vllm/engine/arg_utils.py (hunk 4 above)
Comment (aditi-amd) [reply vllm/engine/arg_utils.py]: I reverted it, it was not needed since we don't have any special_values anyways.

File: vllm/platforms/rocm.py [hunk 5]
```diff
@@ -630,20 +630,27 @@ def get_attn_backend_cls(
...
+                )
+            ):
                 raise ValueError(
                     f"Selected backend {selected_backend} is not valid for "
                     f"this configuration. Reason: {sel_invalid_reasons}"
                 )
             # NOTE: pass a str (not the list) -- info_once hashes its args.
             logger.info_once(
-                "Selected backend %s is incompatible with this turboquant "
+                "Selected backend %s is incompatible with this compressed-KV "
```
Comment (dllehr-amd) [inline vllm/platforms/rocm.py:653]: maybe try with printing `str(kv_dtype)` as well. so we know the exact quant we tried to use that failed

File: vllm/platforms/xpu.py [hunk 6]
```diff
@@ -148,6 +148,10 @@ def get_attn_backend_cls(
     ) -> str:
         # TurboQuant KV cache: route directly to TQ backend
         kv_cache_dtype = attn_selector_config.kv_cache_dtype
+        if kv_cache_dtype == "ultraquant_4bit":
```
Comment (dllehr-amd) [inline vllm/platforms/xpu.py:151]: can we just remove this, probably don't need an explicit check

**Comment by dllehr-amd** [comment]
@AndreasKaratzas  do you have thoughts on a testgroup we can put these flydsl kernel UT's under?  Especially so we can catch api changes when we bump aiter?

File: vllm/v1/attention/backends/turboquant_attn.py [hunk 7]
```diff
@@ -135,12 +139,19 @@ class TurboQuantAttentionBackend(AttentionBackend):
...
         "turboquant_k3v4_nc",
         "turboquant_3bit_nc",
+        "ultraquant_4bit",
     ]
 
     @classmethod
     def customize_spec(cls, spec: AttentionSpec) -> AttentionSpec:
         """TurboQuant packs K+V into one slot per head."""
-        if spec.state_content_bytes is not None or not spec.kv_quant_mode.is_turboquant:
+        if spec.state_content_bytes is not None:
```
Comment (dllehr-amd) [inline vllm/v1/attention/backends/turboquant_attn.py:148]: can we clean this up so we're only dealing with two if statements instead of the three here

**Comment by AndreasKaratzas** [comment]
> @AndreasKaratzas do you have thoughts on a testgroup we can put these flydsl kernel UT's under? Especially so we can catch api changes when we bump aiter?

@dllehr-amd I have:
- https://github.com/vllm-project/vllm/pull/50519

That is going to mirror the Miscellaneous Kernels test group on MI355 to close the parity gap with B200 and that should cover the new tests under `tests/kernels/ultraquant/`.

And for `tests/quantization/test_turboquant.py`, I think it is already covered by `":amd: (MI355) Quantization Shard %N"`. What I would say though is attach a print screen where the tests run and passed as they are @aditi-amd as well as their time to completion.

File: tests/kernels/ultraquant/test_dequant_vs_reference.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+)
+
+
+@pytest.mark.parametrize("dtype", [torch.bfloat16, torch.float16])
+@pytest.mark.parametrize("head_dim", [128, 256])
+@pytest.mark.parametrize("num_kv_heads", [1, 4])
+@pytest.mark.parametrize("cached_len", [32, 96])
+def test_full_dequant_matches_reference(dtype, head_dim, num_kv_heads, cached_len):
+    device = torch.device("cuda")
+    torch.manual_seed(0xD4)
```
Comment (AndreasKaratzas) [inline tests/kernels/ultraquant/test_dequant_vs_reference.py:36]: Can you do a manual review of the tests @aditi-amd please? Spots like `torch.manual_seed(0xD4)` are a bit awkward 😅 
Specifically for this one as well there is a util to set manual seed in vllm instead of using torch api for this. But regardless, check the tests please one by one yourself. I only had a quick glance but I feel like some of them are not really adding any value.

File: vllm/v1/attention/backends/ultraquant_attn.py [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+        attn_metadata: TurboQuantMetadata,
+        Pi: torch.Tensor,
+        centroids: torch.Tensor,
+        PiT: torch.Tensor | None = None,
+        layer: torch.nn.Module | None = None,
+    ) -> torch.Tensor:
+        _mid_o, output_buf, _lse = self._acquire_decode_buffers(query)
+
+        gqa = self.num_kv_groups
+        flydsl_eligible = ultraquant_flydsl_decode_eligible(
```
Comment (dllehr-amd) [inline vllm/v1/attention/backends/ultraquant_attn.py:112]: can we see if there's another place we can make the check if decode is flydsl eligible so we aren't doing this check on every `_decode_attention`

File: vllm/v1/attention/backends/turboquant_attn.py (hunk 3 above)
Comment (dllehr-amd) [reply vllm/v1/attention/backends/turboquant_attn.py:201]: @aditi-amd   can you ensure this is only true for the usecase you care about.  This would enable sinks for all turboquant and ultraquant

File: vllm/v1/attention/backends/turboquant_attn.py
Comment (dllehr-amd) [inline vllm/v1/attention/backends/turboquant_attn.py:127]: Can we just extend `TurboQuantAttentionBackend` into `UltraQuantAttentionBackend` ?  This would let us keep what you want without needing to put ultraquant knowledge into the turboquant code

Comment (aditi-amd) [reply vllm/v1/attention/backends/turboquant_attn.py]: Done, UltraQuantAttentionBackend now subclasses TurboQuantAttentionBackend and UltraQuantAttentionImpl subclasses TurboQuantAttentionImpl; turboquant_attn.py is byte-identical to main now

File: vllm/v1/attention/backends/turboquant_attn.py (hunk 7 above)
Comment (aditi-amd) [reply vllm/v1/attention/backends/turboquant_attn.py]: That edit is reverted; turboquant_attn.py has zero diff against main

File: vllm/v1/attention/backends/turboquant_attn.py (hunk 3 above)
Comment (aditi-amd) [reply vllm/v1/attention/backends/turboquant_attn.py]: Reverted in TurboQuant; sinks support now lives only in UltraQuantAttentionBackend.supports_sink(), where the FlyDSL path is marked ineligible and the Triton fallback handles them.

File: vllm/platforms/xpu.py (hunk 6 above)
Comment (aditi-amd) [reply vllm/platforms/xpu.py]: Removed, xpu.py is no longer touched by this PR

File: vllm/v1/attention/backends/ultraquant_attn.py (hunk 9 above)
Comment (aditi-amd) [reply vllm/v1/attention/backends/ultraquant_attn.py]: Updated, self._flydsl_decode_eligible is computed once in __init__  and _decode_attention only reads the cached flag.

File: vllm/platforms/rocm.py (hunk 5 above)
Comment (aditi-amd) [reply vllm/platforms/rocm.py]: Fixed

File: tests/kernels/ultraquant/test_dequant_vs_reference.py (hunk 8 above)
Comment (aditi-amd) [reply tests/kernels/ultraquant/test_dequant_vs_reference.py]: Fixed, and removed a few redundant tests. torch.manual_seed is gone, all tests now use vllm.utils.torch_utils.set_random_seed

Comment (aditi-amd) [reply tests/kernels/ultraquant/test_dequant_vs_reference.py]: image
Time taken for all the tests and status
